File: typescript/src/types/binary-serializer-datetime.ts
import { DataReader } from "../buffers/data-reader.interface";
import { DataWriter } from "../buffers/data-writer.interface";
import { BinarySerializerLong } from "./binary-serializer-long";
import { BinaryTypeSerializer } from "./binary-type-serializer.interface";

/**
 * Reads (and writes) datetime values to (and from) a binary data source.
 *
 * @public
 * @exported
 * @implements {BinaryTypeSerializer<Date>}
 */
export class BinarySerializerDateTime implements BinaryTypeSerializer<Date> {
    private binarySerializerLong: BinarySerializerLong;

    constructor() {
        this.binarySerializerLong = new BinarySerializerLong();
    }
    
    get sizeInBytes(): number {
        return 8;
    }

    encode(writer: DataWriter, value: Date): void {
        const millisecondsSinceEpoch = this.getMillisecondsSinceEpoch(value);

        this.binarySerializerLong.encode(writer, BigInt(millisecondsSinceEpoch));
    }
    
    decode(reader: DataReader): Date {
        const millisecondsSinceEpoch: bigint = this.binarySerializerLong.decode(reader);

        return new Date(Number(millisecondsSinceEpoch));
    }
    getEquals(a: Date, b: Date): boolean {
        return a.getTime() === b.getTime();
    }

    private getMillisecondsSinceEpoch(value: Date): number {
        return value.getTime();
    }
}

File: dotnet/Barchart.BinarySerializer/Types/BinarySerializerDateTime.cs
#region Using Statements

using Barchart.BinarySerializer.Buffers;
using Barchart.Common.Extensions;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) DateTime values to (and from) a binary data source.
/// </summary>
public class BinarySerializerDateTime : IBinaryTypeSerializer<DateTime>
{
    #region Fields

    private readonly BinarySerializerLong _binarySerializerLong = new();

    #endregion

    #region Methods

    /// <inheritdoc />
    public void Encode(IDataBufferWriter writer, DateTime value)
    {
        _binarySerializerLong.Encode(writer, value.GetMillisecondsSinceUnixEpoch());
    }

    /// <inheritdoc />
    public DateTime Decode(IDataBufferReader reader)
    {
        long millisecondsSinceEpoch = _binarySerializerLong.Decode(reader);

        return DateTime.UnixEpoch.AddMilliseconds(millisecondsSinceEpoch);
    }

    /// <inheritdoc />
    public bool GetEquals(DateTime a, DateTime b)
    {
        return a.Equals(b);
    }

    #endregion
}